### PR TITLE
Sprint #2

### PR DESCRIPTION
**Pull request by fastndead**

### CODE REVIEW

File: client/src/action-creators/cart-action-creator.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import * as cartActions from '../actions/cart-actions';
```
Comment (vadeneev) [inline client/src/action-creators/cart-action-creator.js]: Think about creating Store **folder** for all redux's stuff:
Store
   -cart
       --cart-actions
       --cart-reducer

File: client/src/actions/types.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+export const ERROR_404 = '404';
+
+export const CART_ACTIONS = {
```
Comment (vadeneev) [inline client/src/actions/types.js]: don't mix them, split store by type

File: client/src/assets/basket-linear.svg [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (vadeneev) [inline client/src/assets/basket-linear.svg]: Optimize it

File: client/src/assets/controls.svg [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (vadeneev) [inline client/src/assets/controls.svg]: Optimize

File: client/src/assets/delete.svg [hunk 5]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (vadeneev) [inline client/src/assets/delete.svg]: Optimize

File: client/src/assets/shopping-basket.svg [hunk 6]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (vadeneev) [inline client/src/assets/shopping-basket.svg]: Optimize

File: client/src/components/BestSalesBlock/index.js [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
-  }
-
   render() {
     return (
       <section className='best-sales-block'>
         <div className='best-sales-block__title'>
-          <SectionHeader title_colored='Best' title='sales' description='Lorem Ipsum is simply dummy text of the printing and typesetting industry' />
+          <Translation>
+            {
+              t => <SectionHeader title_colored={t('bestSales.best')} title={t('bestSales.sales')} description={t('bestSales.description')} />
```
Comment (vadeneev) [inline client/src/components/BestSalesBlock/index.js]: It looks bad, that's why I'm suggesting to use i18n without react wrapper

File: client/src/components/CartWindow/CartWindow.jsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import OrderBlock from './OrderBlock/OrderBlock';
+
+import './CartWindow.scss';
+
+export function CartWindow() {
+  return (
+    <div className="card-window">
+      <button className="card-window__background" onClick={() => store.dispatch(closeCart())}></button>
+      <div className="card-window__content">
+        <CloseButton onClick={() => store.dispatch(closeCart())}/>
```
Comment (vadeneev) [inline client/src/components/CartWindow/CartWindow.jsx]: no, we should use connect + mapDispatchToProps

File: client/src/components/CartWindow/Items/Item/Item.jsx [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  }
+
+  handleOnClickDecrement(event) {
+    if (Number(this.props.data.amount) > 1) {
+      store.dispatch(updateItem(store.getState(), this.props.data, this.props.data.color, this.props.data.size, this.props.data.amount - 1));
+    }
+  }
+
+  render() {
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Item/Item.jsx]: I have no idea what data used to render this component, no one knows, except person who wrote this component

File: client/src/components/CartWindow/Items/Item/Item.jsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import store from '../../../../store';
+import removeItem from '../../../../utils/cart/removeItem';
+import updateItem from '../../../../utils/cart/updateItem';
+
+import './Item.scss';
+
+export class Item extends React.Component {
+  handleOnClickIncrement(event) {
+    if (Number(this.props.data.amount) < 99) {
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Item/Item.jsx]: Why amount is string ?

File: client/src/components/CartWindow/Items/Item/Item.jsx [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+      store.dispatch(updateItem(store.getState(), this.props.data, this.props.data.color, this.props.data.size, this.props.data.amount + 1));
+    }
+  }
+
+  handleOnClickDecrement(event) {
+    if (Number(this.props.data.amount) > 1) {
+      store.dispatch(updateItem(store.getState(), this.props.data, this.props.data.color, this.props.data.size, this.props.data.amount - 1));
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Item/Item.jsx]: ??!?

File: client/src/components/CartWindow/Items/Item/Item.jsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+              Size:
+              <span className="cart-window-item__size"> {this.props.data.size}</span>
+            </p>
+            <div className="cart-window-item__color-wrapper">
+              <p className="cart-window-item__color-text">Color: </p>
+              <div style={{ background: this.props.data.color }} className="cart-window-item__color"></div>
+            </div>
+            <p className="cart-window-item__price-wrapper">
+              Price:
+              <span className="cart-window-item__price"> {this.props.data.generalData.price}$</span>
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Item/Item.jsx]: How about rubles ?

File: client/src/components/CartWindow/Items/Items.jsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+
+import './Items.scss';
+import { connect } from 'react-redux';
+import { Item } from './Item/Item';
+
+function Items(props) {
+  const items = props.items.map((item, index) => {
+    return <Item
+            key={index}
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Items.jsx]: we should not use index as a key

File: client/src/components/CartWindow/Items/Items.jsx [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+            index={index}
+            data={item}
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Items.jsx]: component's should not be dependent from complex types

File: client/src/components/CartWindow/Items/Items.scss [hunk 15]
```diff
@@ -0,0 +1,23 @@
+.card-window {
+  &__items {
+    flex-grow: 1;
+    overflow-y: auto;
+    box-sizing: border-box;
+    max-height: calc(100% - 50px);
+    padding: 40px 50px 0 50px;
+    background: #ffffff;
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Items.scss]: use variables

File: client/src/components/CartWindow/Items/Items.scss [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+      width: 3px;
+    }
+     
+    &::-webkit-scrollbar-track {
+      background: transparent;
+    }
+     
+    &::-webkit-scrollbar-thumb {
+      border-radius: 4px;
+      background: #ff5912;
```
Comment (vadeneev) [inline client/src/components/CartWindow/Items/Items.scss]: use variables

File: client/src/components/CartWindow/OrderBlock/OrderBlock.jsx [hunk 17]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import { connect } from 'react-redux';
+
+import './OrderBlock.scss';
+
+function OrderBlock(props) {
+  return (
+    <div className="card-window__order">
+      <p className="card-window__order-price-wrapper">
+        Order price: <span className="card-window__order-price">{`${props.orderPrice || 0}$`}</span>
```
Comment (vadeneev) [inline client/src/components/CartWindow/OrderBlock/OrderBlock.jsx]: Rubles, euro ?

File: client/src/components/CartWindow/OrderBlock/OrderBlock.scss [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  z-index: 5;
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  height: 100px;
+  background: #ffffff;
+  box-shadow: 0px 2px 17px 0px rgba(0,0,0,0.15);
+
+  &-price{
+    color: #ff5912;
```
Comment (vadeneev) [inline client/src/components/CartWindow/OrderBlock/OrderBlock.scss]: use variables

File: client/src/components/CartWindow/OrderBlock/OrderBlock.scss [hunk 19]
```diff
@@ -0,0 +1,43 @@
+@import "../../../style/main.scss";
+
+.card-window__order {
+  z-index: 5;
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  height: 100px;
+  background: #ffffff;
```
Comment (vadeneev) [inline client/src/components/CartWindow/OrderBlock/OrderBlock.scss]: use variables

File: client/src/components/Filters/FIlterSlider/FilterSlider.js [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+import {toast} from 'react-toastify';
+import store from "../../../store";
+import {changeMinPriceFilter, changeMaxPriceFilter} from "../../../action-creators/filter-action-creator";
+
+
+class FilterSlider extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      maxValue: props.maxValue,
```
Comment (vadeneev) [inline client/src/components/Filters/FIlterSlider/FilterSlider.js]: it is not recommended to initiate state with props

File: client/src/components/Filters/FIlterSlider/FilterSlider.js [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+
+  handleChange = (chosenValues) => {
+    const [minChosenValue, maxChosenValue] = chosenValues;
+    this.setState({minChosenValue, maxChosenValue})
+  };
+
+  handleInputChange = (e) => {
+    const value = Number(e.target.value.replace(/\D/g, ''));
+    if (Number.isInteger(value)) {
+      switch (e.target.name) {
```
Comment (vadeneev) [inline client/src/components/Filters/FIlterSlider/FilterSlider.js]: don't waste component with such structures, move it to helper, and return value to set state

File: client/src/components/Filters/FIlterSlider/FilterSlider.js [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+  componentDidUpdate(prevProps, prevState, snapshot) {
+    if(prevState.maxChosenValue !== this.state.maxChosenValue) {
+      store.dispatch(changeMaxPriceFilter( this.state.maxChosenValue));
+    }
+    if(prevState.minChosenValue !== this.state.minChosenValue) {
+      store.dispatch(changeMinPriceFilter( this.state.minChosenValue));
+    }
+  }
+
+  render() {
```
Comment (vadeneev) [inline client/src/components/Filters/FIlterSlider/FilterSlider.js]: spread data before passing it - const {a,b,c} = this.state

Comment (vadeneev) [reply client/src/components/Filters/FIlterSlider/FilterSlider.js]: it is handy

File: client/src/components/Filters/Filters.js [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+    this.state = {
+      filters: null
+    }
+  }
+
+  componentDidMount() {
+    loadFilters().then(res => {
+          this.setState({
+            filters: res.data,
+          })
```
Comment (vadeneev) [inline client/src/components/Filters/Filters.js]: looks like a stairs to hell

File: client/src/utils/notificationSuccess.js [hunk 24]
```diff
@@ -0,0 +1,11 @@
+import { toast } from 'react-toastify';
+
+import i18n from '../i18n';
+
+export default function notificationSuccess(messageRu, messageEn, name) {
```
Comment (vadeneev) [inline client/src/utils/notificationSuccess.js]: default export is not your friend

File: client/src/utils/notificationError.js [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+import i18n from '../i18n';
+
+export default function notificationError(messageRu, messageEn, error) {
+  let errorCode = error.message.match(/\d+/)[0];
+  if (errorCode >= 400 && errorCode < 500) {
+    if (i18n.language === 'ru') {
+      (() => toast(messageRu, { type: toast.TYPE.ERROR }))();
+    } else {
+      (() => toast(messageEn, { type: toast.TYPE.ERROR }))();
```
Comment (vadeneev) [inline client/src/utils/notificationError.js]: so much code duplicates, only one difference with 9th line is message var

File: client/src/utils/cart/removeItem.js [hunk 26]
```diff
@@ -0,0 +1,27 @@
+import { updateItems as updateItemsActionCreator } from '../../action-creators/cart-action-creator';
+
+import updateLocalStorageCollection from '../localStorage/updateLocalStorageCollection';
+import notificationSuccess from '../notificationSuccess';
+
+export default function removeItem(state, target) {
```
Comment (vadeneev) [inline client/src/utils/cart/removeItem.js]: Take a look at redux-thunk or saga

File: client/src/reducers/cartReducer.js [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+        let orderPrice = 0;
+        for (const item of action.items) {
+          orderPrice += item.generalData.price * item.amount;
+        }
+        return {
+          ...state,
+          items: action.items,
+          size: action.items.length,
+          orderPrice,
+          opened: state.opened && !state.loading && !state.failure && action.items.length
```
Comment (vadeneev) [inline client/src/reducers/cartReducer.js]: it is a derived data, not state

File: client/src/index.scss [hunk 28]
```diff
@@ -2,3 +2,13 @@ body {
   margin: 0;
   padding: 0;
 }
+
+.scroll-disabled {
+  height: 100%;
+  overflow: hidden;
```
Comment (vadeneev) [inline client/src/index.scss]: It wouldn't work well

File: client/src/components/ProductSmallContainer/index.js [hunk 29]
```diff
@@ -11,16 +13,26 @@ export class ProductSmallContainer extends Component {
...
+  componentDidMount() {
+    loadCard(1, 3).then(result => {
+      this.setState({
+        ready: true,
+        cards: [...this.state.cards, ...result.data.items],
+        loading: false,
+        nextPage: result.data.nextPage
+      })
+    }).catch((error) => {
+      notificationError('Таких товаров не существует', 'Products you\'re looking for is nowhere to be found.', error)
```
Comment (vadeneev) [inline client/src/components/ProductSmallContainer/index.js]: don't use hardcoded values in components, at least - move them to constants file

File: client/src/components/Product/ProductImages/index.js [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+  render() {
+    return (
+        <div className="product-images">
+          <img className="product-images__main" src={this.state.images[this.state.currentIndex]}
+               alt="Product"/>
+
+          <div className="product-images__row">
+            {this.state.images.map((image, index) => {
+              return <div key={index} onClick={() => this.setIndex(index)}
+                          className={`product-images__image-wrapper
```
Comment (vadeneev) [inline client/src/components/Product/ProductImages/index.js]: take a look at Classnames

File: client/src/components/Filters/FIlterSlider/FilterSlider.js (hunk 20 above)
Comment (fastndead) [reply client/src/components/Filters/FIlterSlider/FilterSlider.js]: this props in particular are only seed data, component will not receive this props in the future, this state value will be only controlled inside the component, put props needed to initialize some data without hardcoding it inside the component, which will be not useful in terms of reusability of this component

**Pull request merged**